### PR TITLE
pavics compose.sh: overriding `DATA_PERSIST_ROOT` in `env.local` do not take effect and warn when a dir in `EXTRA_CONF_DIRS` does not exist

### PR DESCRIPTION
**Pull request by tlvu**
## Fixes:

- Overriding `DATA_PERSIST_ROOT` in `env.local` do not take effect for
  `JUPYTERHUB_USER_DATA_DIR`, `MAGPIE_PERSIST_DIR`, and `GEOSERVER_DATA_DIR`.

  These 3 vars will have to be delayed evaluated for override in `env.local` to
  take effect.

  For a variable to be delayed evaluated, it has to be defined using
  single-quote and be added to the list of `DELAYED_EVAL` in `default.env`.

  If those steps are forgotten in `env.local`, it will still work since
  `env.local` is the last file to be read.  However those steps should not be
  forgotten in any `default.env` for all components.

  So the impact or burden is on the developpers to write their `default.env`
  file properly, not on the users that only modify the `env.local` file.

  All `default.env` files header have been updated with notice about this new
  delayed evaluation feature.

  Fixes https://github.com/bird-house/birdhouse-deploy/issues/270

## Changes:

- Warn when a dir in `EXTRA_CONF_DIRS` does not exist.

  Most likely a typo in a new dir.  Just warn and not exit directly to avoid
  leaving the entire platform down during an unattended autodeploy since no
  one is around to take immediate action.

  Fixes https://github.com/bird-house/birdhouse-deploy/issues/266

### CODE REVIEW

**Comment by tlvu** [comment]
@fmigneault I changed my mind and go ahead with single-quote and delayed evaluation instead.

The impact on users is actually minimum since forgetting to use single-quote in `env.local` has no problem.  Basically the complexity burden is on the developer only.  I've believe I've left enough documentation comments for the developers.

File: birdhouse/default.env [hunk 1]
```diff
@@ -25,7 +27,8 @@ export BASH_IMAGE="bash:5.1.4"
 export DATA_PERSIST_ROOT="/data"
 
 # Folder on the host to persist Jupyter user data (noteboooks, HOME settings)
-export JUPYTERHUB_USER_DATA_DIR="${DATA_PERSIST_ROOT}/jupyterhub_user_data"
+# Must use single-quote for delayed eval.
+export JUPYTERHUB_USER_DATA_DIR='$DATA_PERSIST_ROOT/jupyterhub_user_data'
```
Comment (fmigneault) [inline birdhouse/default.env:31]: Can we leave the `{}` for better readability (here and everywhere else where quotes were changed)? It is easier to spot variables with them (`$` by itself is too easily confused with `S` depending on font style/size).

File: CHANGES.md [hunk 2]
```diff
@@ -14,7 +14,39 @@
...
+  file properly, not on the users that only modify the `env.local` file.
+
+  All `default.env` files header have been updated with notice about this new
+  delayed evaluation feature.
+
+  Fixes https://github.com/bird-house/birdhouse-deploy/issues/270.
+
+## Changes:
+
+- Warn when a dir in `EXTRA_CONF_DIRS` do not exist.
```
Comment (fmigneault) [inline CHANGES.md:42]: typo "does"

File: CHANGES.md [hunk 3]
```diff
@@ -14,7 +14,39 @@
...
+
+  Most likely a typo in a new dir.  Just warn and not exit directly to avoid
+  leaving the entire platform down during an unattended autodeploy since no
+  one around to take immediate action.
```
Comment (fmigneault) [inline CHANGES.md:46]: typo "is around"

File: birdhouse/pavics-compose.sh [hunk 4]
```diff
@@ -95,6 +95,13 @@ COMPOSE_DIR="`pwd`"
 [ -f env.local ] && . ./env.local
 
 for adir in ${EXTRA_CONF_DIRS}; do
+  if [ ! -e "$adir" ]; then
+    # Do not exit to not break unattended autodeploy since no human around to
+    # fix immediately.
+    # The new adir with typo will not be active but at least all the existing
+    # will still work.
+    echo "WARNING: EXTRA_CONF_DIRS '$adir' do not exist" 1>&2
```
Comment (fmigneault) [inline birdhouse/pavics-compose.sh:103]: Same "does" typo.

Maybe reword also because is it one item of `EXTRA_CONF_DIRS` that is flagged here.
Current wording could be misinterpreted as `EXTRA_CONF_DIRS = $adir` (*only*).

File: birdhouse/pavics-compose.sh [hunk 5]
```diff
@@ -106,6 +113,12 @@ done
 # components.
 [ -f env.local ] && . ./env.local
 
+for i in ${DELAYED_EVAL}; do
+  v="`eval "echo \\$${i}"`"
+  eval 'export ${i}="`eval "echo ${v}"`"'
+  echo "delayed eval '`env |grep ${i}=`'"
+done
```
Comment (fmigneault) [inline birdhouse/pavics-compose.sh:120]: I'm not sure if I understand the need for the separate `DELAYED_EVAL` definition. 
Wouldn't developers forgetting to specify a new variable name in it be more prone to error? 
Couldn't all variables in `${VARS}` could do the ```v="`eval "echo \\$${i}"`"``` step, and this way simply using the single quotes would be sufficient to achieve the same result?

File: birdhouse/default.env (hunk 1 above)
Comment (tlvu) [reply birdhouse/default.env:31]: done

File: CHANGES.md (hunk 2 above)
Comment (tlvu) [reply CHANGES.md:42]: done

File: birdhouse/pavics-compose.sh (hunk 4 above)
Comment (tlvu) [reply birdhouse/pavics-compose.sh:103]: done

File: CHANGES.md (hunk 3 above)
Comment (tlvu) [reply CHANGES.md:46]: done

File: birdhouse/pavics-compose.sh (hunk 5 above)
Comment (tlvu) [reply birdhouse/pavics-compose.sh:120]: Well, if you think for simplicity, all vars in `default.env` should be delayed eval then `VARS` do not contain all of them.  `VARS` currently do not contain the 3 broken vars in `DELAYED_EVAL`.

Then it looks like delayed eval is an exception rather than the norm.  So I'd rather not forced the single quote on more vars than needed.

File: birdhouse/default.env [hunk 6]
```diff
@@ -1,5 +1,7 @@
 # All env in this default.env can be overridden by env.local.
-# All env in this default.env must not depend on any env in env.local.
+
+# All env in this default.env must NOT depend on any other env.  If they do,
+# must use single quote and must add to the list of DELAYED_EVAL.
```
Comment (fmigneault) [inline birdhouse/default.env:4]: I think it would be worth adding the reason here that "[...] must use single quote *to avoid early expansion before overrides in local.env are applied* [...]".

File: birdhouse/pavics-compose.sh (hunk 5 above)
Comment (tlvu) [reply birdhouse/pavics-compose.sh:120]: Ouch, the grep trick might too magic for me.  How do you get the name of each var after?

File: birdhouse/default.env (hunk 6 above)
Comment (tlvu) [reply birdhouse/default.env]: done in 407c348e576b917657bc8b2de1e3b5e5f8c45fe5

File: birdhouse/pavics-compose.sh (hunk 5 above)
Comment (fmigneault) [reply birdhouse/pavics-compose.sh:120]: Using this:
 cat birdhouse/default.env | grep -E "export [A-Z0-9_]+=" | cut -d ' ' -f 2 | cut -d '=' -f 1`

Comment (fmigneault) [reply birdhouse/pavics-compose.sh:120]: And an extra ` | sed -r -e 's/(.*)/\$\1/g'` at the end for adding the `$`.

Comment (tlvu) [reply birdhouse/pavics-compose.sh:120]: Oh wow, think I'll pass on this magic for now.  We have to handle all the `default.env` from components from any other external repos as well.  I don't have time to think of all the corner cases with this magic for now.

Do you see anything blocking with the current approach?

Comment (fmigneault) [reply birdhouse/pavics-compose.sh:120]: No. Good to go with the current alternative.

**Review by fmigneault — APPROVED** [review]
Current method sufficient for now. 
Alternative to avoid `DELAYED_EVAL` to be considered as some other time as per https://github.com/bird-house/birdhouse-deploy/pull/272#discussion_r1055829225.

For reference: https://github.com/bird-house/birdhouse-deploy/issues/274
